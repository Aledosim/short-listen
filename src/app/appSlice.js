import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  value: 'trackList',
}

export const appSlice = createSlice({
  name: 'app',
  initialState,
  reducers: {
    changedToFavorites: (state) => {
      console.log('mudando para favoritos')
      state.value = 'favorites'},
  },
})

export const { changedToFavorites } = appSlice.actions

export const selectView = (state) => state.app.value

export default appSlice.reducer
