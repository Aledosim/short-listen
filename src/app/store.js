import { configureStore } from '@reduxjs/toolkit';

import trackListReducer from '../features/tracklist/trackListSlice'
import favoritesReducer from '../features/favorites/favoritesSlice'

export const store = configureStore({
  reducer: {
    trackList: trackListReducer,
    favorites: favoritesReducer,
  },
});
