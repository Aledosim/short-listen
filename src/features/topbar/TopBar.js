import React, { useContext } from 'react'

export default function TopBar(){

  return(
    <header data-cy='topbar'>
      <form
        onSubmit={() => {}}
      >
        <input
          type='search'
          autoComplete='on'
          autoSave='true'
          placeholder='Album, artist or title'
          id='searchField'
          data-cy='searchField'
        />
        <button
          type='submit'
          data-cy='searchButton'
        />
      </form>
    </header>
  );
};
