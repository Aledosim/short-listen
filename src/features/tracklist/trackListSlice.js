import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'

import { fetchTracksChart, fetchTerm } from '../../utils/apiclient'

const initialState = {
  value: [],
  status: 'idle',  // idle, loading, suceeded, failed
  error: null,
  searchTerm: '',
  isSearch: false,
}

export async function fetchTracksChartThunk({index, limit} = {index: undefined, limit: undefined}) {

  const tracksChart = await fetchTracksChart({index: index, limit: limit})

  return tracksChart.data.tracks.data
}

export const trackChartEnded = createAsyncThunk(
  'trackList/trackChartEnded',
  fetchTracksChartThunk,
  {
    condition: (arg, { getState }) => {
      const { trackList } = getState()
      const fetchStatus = trackList.status
      if (fetchStatus === 'loading'){
        return false
      }
    }
  }
)

export async function fetchTermThunk({index, limit} = {index: undefined, limit: undefined}, thunkAPI) {
  const term = thunkAPI.getState().trackList.searchTerm
  const searchResult = await fetchTerm(term, {index, limit})

  // thunkAPI.dispatch(resetedTrackList())
  return searchResult.data.data
}

)

export const trackListSlice = createSlice({
  name: 'trackList',
  initialState,
  reducers: {
    searchedTerm: (state, action) => {
      state.searchTerm = action.payload
      state.isSearch = true
    }
  },
  // The `extraReducers` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers: {
    [trackChartEnded.pending]: (state) => {state.status = 'loading'},
    [trackChartEnded.fulfilled]: (state, action) => {

      const chartTracks = action.payload
      const newTracks = chartTracks.map( track => {
        return {
          cover: track.album.cover,
          id: track.id,
          link: track.link,
          preview: track.preview,
          singer: track.artist.name,
          time: track.duration,
          title: track.title
        }
      })

      state.status = 'suceeded'
      state.value = state.value.concat(newTracks)
    }
  },
})

export const {searchedTerm} = trackListSlice.actions

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state: RootState) => state.counter.value)`
export const selectTrackList = (state) => state.trackList.value

export default trackListSlice.reducer
