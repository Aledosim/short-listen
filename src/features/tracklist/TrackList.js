import React, { useEffect } from 'react'
import styled from 'styled-components'
import InfiniteScroll from 'react-infinite-scroll-component'
import { useSelector, useDispatch } from 'react-redux'

import { trackChartEnded, selectTrackList } from './trackListSlice'
import TrackCard from '../trackcard/TrackCard'

// colors
const text = "#EBEBEB"

const StyledInfiniteScroll = styled(InfiniteScroll)`
  display: flex;
  flex-direction: column;
  align-items: center;

  color: ${text};
`

export default function TrackList(){

  const trackList = useSelector(selectTrackList)
  const dispatch = useDispatch()

  const pageSize = 30
  var isSearch = false

  useEffect(() => {
    dispatch(trackChartEnded({index: 0, limit: pageSize}))
  }, [dispatch])

  function fetchMoreData(){
    console.log("fetchMoreData")
    dispatch(trackChartEnded({index: trackList.length, limit: pageSize}))
  }

  return(
    <div role="list">
      <StyledInfiniteScroll
      dataLength={trackList.length} //This is important field to render the next data
      next={fetchMoreData}
      hasMore={true}
      loader={<h4>Loading...</h4>}
      endMessage={
        <p style={{ textAlign: 'center' }}>
        <b>Yay! You have seen it all</b>
        </p>
      }
      >
      {
        trackList.map( (track, i) => {
          return(
              <TrackCard
                cover={track.cover}
                title={track.title}
                singer={track.singer}
                time={track.time}
                preview={track.preview}
                link={track.link}
                id={track.id}
                key={i}
              />
          )
        })
      }
      </StyledInfiniteScroll>
    </div>
  );
};

