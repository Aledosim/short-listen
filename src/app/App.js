import React from 'react';
import style, { createGlobalStyle } from 'styled-components';

import TopBar from '../features/topbar/TopBar'
import TrackList from '../features/tracklist/TrackList'

// colors
const background = "#0A0A0A"

const GlobalStyle = createGlobalStyle`
	* {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
	}

  body {
    background-color: ${background};
  }

	@media(max-width: 1080px) {
    html {
      font-size: 93.75%;
    }
	}

	@media(max-width: 720px) {
    html {
      font-size: 87.5%;
    }
	}

  button {
    cursor: pointer;
  }
`

const Header = style.header`
  position: fixed;
  width: 100%;
  z-index: 1;

  background-color: ${background};
`

const TrackListContainer = style.div`
  position: relative;
  top: 4.5rem;
`

function App() {
  return (
    <>
      <GlobalStyle />
      <Header>
        <TopBar />
      </Header>
      <TrackListContainer>
        <TrackList />
      </TrackListContainer>
    </>
  );
}

export default App;
