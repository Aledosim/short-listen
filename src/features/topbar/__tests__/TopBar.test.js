import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { Provider } from 'react-redux'

import { store } from '../../../app/store'
import TopBar, { search } from '../TopBar'
import * as TopBarModule from '../TopBar'
import * as trackListSlice from '../../tracklist/trackListSlice'

function testRender() {
    return render(
      <Provider store={store}>
        <TopBar />
      </Provider>
    )
}

describe('search tests', () => {

  it('returns a function', () => {
    const func = search(jest.fn())

    expect(typeof func).toEqual('function')
  })

  it('gets the correct input', () => {
    const event = { preventDefault: () => {} }

    document.getElementById = jest.fn(() => HTMLInputElement)
    HTMLInputElement.value = ''

    const func = search(() => {})
    func(event)

    expect(document.getElementById).toHaveBeenCalledWith('searchField')
  })

  it('search for lowercase', () => {
    const event = { preventDefault: () => {} }

    const dispatchMock = jest.fn()

    HTMLInputElement.value = 'SomE CrAzY TeXT'
    document.getElementById = () => HTMLInputElement

    const termSearchedMock = jest.spyOn(trackListSlice, 'searchedTerm')

    const func = search(dispatchMock)
    func(event)

    expect(termSearchedMock).toHaveBeenCalledWith('some crazy text')
    expect(dispatchMock).toHaveBeenCalled()
  })

  it('calls event.disableDefault method to prevent page reload', () => {
    const mockPreventDefault = jest.fn()
    const event = { preventDefault: mockPreventDefault }

    HTMLInputElement.value = ''
    document.getElementById = () => HTMLInputElement

    const func = search(() => {})
    func(event)

    expect(mockPreventDefault).toHaveBeenCalled()
  });
})

describe('<TopBar /> tests', () => {

  it('renders without crashing', async () => {
    const tree = testRender()

    expect(tree).toMatchSnapshot()
  })

  it('call search when search button is clicked', async () => {
    fail("I could not mock this function in time")
    const mockSearch = jest.spyOn(TopBarModule, 'search')

    testRender()

    // The click event flow for forms is not implemented on jsdom
    // https://github.com/jsdom/jsdom/issues/1937
    // fireEvent.submit(screen.getByRole('button'))

    const button = screen.getByRole('button')
    fireEvent.click(button)

    expect(mockSearch).toHaveBeenCalled()
  })
})
