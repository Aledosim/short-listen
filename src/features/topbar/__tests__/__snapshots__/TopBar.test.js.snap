// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopBar /> tests renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  max-height: 50px;
  padding-top: 0.75rem;
  padding-bottom: 0.5rem;
}

.c1 {
  margin-left: 2.5rem;
  max-width: 100px;
  height: auto;
}

.c2 {
  position: relative;
  margin: auto;
  width: 30rem;
  height: 3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  width: 100%;
  height: 80%;
  padding-right: 4rem;
  padding-left: 2rem;
  background-color: #F4F1F1;
  border: none;
  border-radius: 4px;
}

.c3::-webkit-input-placeholder {
  color: #000000,46%;
}

.c3::-moz-placeholder {
  color: #000000,46%;
}

.c3:-ms-input-placeholder {
  color: #000000,46%;
}

.c3::placeholder {
  color: #000000,46%;
}

.c4 {
  position: relative;
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
  right: 9px;
  height: 70%;
  width: 3rem;
  border: none;
  background-image: url(search_icon.svg);
  background-repeat: no-repeat;
  background-color: #F4F1F1;
  color: #9B9B9B;
}

<body>
    <div>
      <header
        class="c0"
        data-cy="topbar"
      >
        <svg
          class="c1"
        >
          deezer_logo.svg
        </svg>
        <form
          class="c2"
        >
          <input
            autocomplete="on"
            autosave="true"
            class="c3"
            data-cy="searchField"
            id="searchField"
            placeholder="Album, artist or track title"
            type="search"
          />
          <button
            class="c4"
            data-cy="searchButton"
            type="button"
          />
        </form>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="sc-bdnxRM dtpuxP"
      data-cy="topbar"
    >
      <svg
        class="sc-gtsrHT cBzwgK"
      >
        deezer_logo.svg
      </svg>
      <form
        class="sc-dlnjwi gSWzLL"
      >
        <input
          autocomplete="on"
          autosave="true"
          class="sc-hKFxyN cdSgUH"
          data-cy="searchField"
          id="searchField"
          placeholder="Album, artist or track title"
          type="search"
        />
        <button
          class="sc-eCApnc kAxBQA"
          data-cy="searchButton"
          type="button"
        />
      </form>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
